### PR TITLE
Spar Polysemy: Remove ReaderT from Spar

### PR DESCRIPTION
**Pull request by isovector**
This PR removes the `ReaderT` inside of the `Spar` newtype, instead using the `Input` effect to thread the `Opts` to whichever places need it.

To be merged after #1814.

## Checklist

 - [x] The **PR Title** explains the impact of the change.
 - [x] The **PR description** provides context as to why the change should occur and what the code contributes to that effect. This could also be a link to a JIRA ticket or a Github issue, if there is one.
 - [x] A file with the changelog entry in one or more suitable sub-sections. The sub-sections are marked by directories inside `changelog.d`.

### CODE REVIEW

File: services/spar/src/Spar/App.hs [hunk 1]
```diff
@@ -724,7 +779,7 @@ verdictHandlerMobile granted denied = \case
 -- | When getting stuck during login finalization, show a nice HTML error rather than the json
 -- blob.  Show lots of debugging info for the customer to paste in any issue they might open.
 errorPage :: SparError -> [Multipart.Input] -> Maybe Text -> ServerError
-errorPage err inputs mcky =
+errorPage err mp_inputs mcky =
```
Comment (fisx) [inline services/spar/src/Spar/App.hs:782]:
```suggestion
errorPage err mpInputs mcky =
```

?

File: services/spar/src/Spar/App.hs [hunk 2]
```diff
@@ -176,8 +173,9 @@ data Env = Env
     sparCtxRequestId :: RequestId
   }
 
-instance HasConfig (Spar r) where
-  getConfig = asks (saml . sparCtxOpts)
+-- TODO(sandy): This is the only use of the Reader effect
```
Comment (fisx) [inline services/spar/src/Spar/App.hs:176]: You mean "should be made", and "`(Input Opts)` effect"?

```
git grep -Hn '\binput\b' 
src/Spar/App.hs:570:      logr <- liftSem input
src/Spar/Scim.hs:166:            logger <- input @TinyLog.Logger
src/Spar/Sem/AReqIDStore/Cassandra.hs:27:    denv <- Data.mkEnv <$> input <*> (fromTime <$> getNow)
src/Spar/Sem/AssIDStore/Cassandra.hs:23:      denv <- Data.mkEnv <$> input <*> (fromTime <$> getNow)
src/Spar/Sem/BindCookieStore/Cassandra.hs:25:    denv <- Data.mkEnv <$> input <*> (fromTime <$> getNow)
```

File: services/spar/src/Spar/App.hs [hunk 3]
```diff
@@ -518,12 +562,13 @@ verdictHandlerResult bindCky mbteam verdict = do
...
 
-catchVerdictErrors :: Spar r VerdictHandlerResult -> Spar r VerdictHandlerResult
+catchVerdictErrors :: forall r. Member (Input TinyLog.Logger) r => Spar r VerdictHandlerResult -> Spar r VerdictHandlerResult
 catchVerdictErrors = (`catchError` hndlr)
   where
     hndlr :: SparError -> Spar r VerdictHandlerResult
     hndlr err = do
-      logr <- asks sparCtxLogger
+      logr <- liftSem input
+      -- TODO(sandy): When we remove this line, we can get rid of the @Input TinyLog.Logger@ effect
```
Comment (fisx) [inline services/spar/src/Spar/App.hs:571]: nice.

**Review by fisx — APPROVED** [review]

File: services/spar/test-integration/Util/Core.hs [hunk 4]
```diff
@@ -1247,60 +1234,38 @@ runSimpleSP action = do
...
-        . runExceptT
-        $ runReaderT action env
+        . defaultSsoCodeToCassandra
+        . scimTokenStoreToCassandra
+        . scimUserTimesStoreToCassandra
+        . scimExternalIdStoreToCassandra
+        . aReqIDStoreToCassandra
+        . assIDStoreToCassandra
+        . bindCookieStoreToCassandra
+        $ runExceptT action
```
[the remark this answers is not in the document]
Comment (isovector) [reply services/spar/test-integration/Util/Core.hs:1268]: That's coming up immediately in the next PR :)

File: services/spar/src/Spar/Sem/AReqIDStore/Cassandra.hs [hunk 5]
```diff
@@ -19,12 +19,12 @@ instance Member (Embed IO) r => HasNow (Sem r)
 
 aReqIDStoreToCassandra ::
   forall m r a.
-  (MonadClient m, Members '[Embed m, Error TTLError, Embed IO, Reader Opts] r) =>
+  (MonadClient m, Members '[Embed m, Error TTLError, Embed IO, Input Opts] r) =>
```
[the remark this answers is not in the document]
Comment (isovector) [reply services/spar/src/Spar/Sem/AReqIDStore/Cassandra.hs:22]: Yeah, but so does the `Reader` effect! My rationale here is that "nobody is using `local` so `Reader` is not actually the effect they want."

File: stack.yaml [hunk 6]
```diff
@@ -175,6 +175,7 @@ extra-deps:
 - servant-swagger-ui-0.3.4.3.36.1
 - tls-1.5.5
 - cryptonite-0.28
+- polysemy-1.6.0.0
```
[the remark this answers is not in the document]
Comment (isovector) [reply stack.yaml:178]: This is an upgrade! But yes.

**Pull request merged**